src/renderer/renderScene.ts

```diff
@@ -322,357 +325,248 @@ const renderLinearElementPointHighlight = (
     hoverPointIndex,
   );
   context.save();
-  context.translate(renderConfig.scrollX, renderConfig.scrollY);
+  context.translate(appState.scrollX, appState.scrollY);
 
-  highlightPoint(point, context, renderConfig);
+  highlightPoint(point, context, appState);
   context.restore();
 };
 
 const frameClip = (
   frame: ExcalidrawFrameElement,
   context: CanvasRenderingContext2D,
-  renderConfig: RenderConfig,
+  renderConfig: StaticCanvasRenderConfig,
+  appState: StaticCanvasAppState,
 ) => {
-  context.translate(
-    frame.x + renderConfig.scrollX,
-    frame.y + renderConfig.scrollY,
-  );
+  context.translate(frame.x + appState.scrollX, frame.y + appState.scrollY);
   context.beginPath();
   if (context.roundRect && !renderConfig.isExporting) {
     context.roundRect(
       0,
       0,
       frame.width,
       frame.height,
-      FRAME_STYLE.radius / renderConfig.zoom.value,
+      FRAME_STYLE.radius / appState.zoom.value,
     );
   } else {
     context.rect(0, 0, frame.width, frame.height);
   }
   context.clip();
   context.translate(
-    -(frame.x + renderConfig.scrollX),
-    -(frame.y + renderConfig.scrollY),
+    -(frame.x + appState.scrollX),
+    -(frame.y + appState.scrollY),
   );
 };
 
-export const _renderScene = ({
-  elements,
-  appState,
-  scale,
-  rc,
+const getNormalizedCanvasDimensions = (
+  canvas: HTMLCanvasElement,
+  scale: number,
+): [number, number] => {
+  // When doing calculations based on canvas width we should used normalized one
+  return [canvas.width / scale, canvas.height / scale];
+};
+
+const bootstrapCanvas = ({
```

Comment:
Didn't review yet, so leaving a note that this helper is doing lots of `context.save()` calls without any corresponding `.restore()` calls, which I assume is the result of cut & paste refactor. We'll want to move these saves outside this helper (and similar helpers if applicable) into the call site's context.

Reply:
removed one unnecessary context.save() right before scaling, moved the zoom with context.save() outside the bootstrapCanvas